File: LeetCode/1510. Stone Game IV/solution.py
from functools import lru_cache
import sys
sys.setrecursionlimit(10**6)

class Solution:
    @lru_cache(None)
    def winnerSquareGame(self, n: int) -> bool:
        if n == 0:
            return False
        i = 1
        while i*i <= n:
            if not self.winnerSquareGame(n-i*i):
                return True
            i += 1
        return False


class Solution:
    def winnerSquareGame(self, n: int) -> bool:
        dp = [False]*(n+1)
        for i in range(1,n+1):
            j = 1
            while j*j <= i:
                if not dp[i-j*j]:
                    dp[i] = True
                    break
                j += 1
        return dp[-1]

File: LeetCode/1510. Stone Game IV/solution.cpp
const int maxn = 1e5+5;

bool dp[maxn];

class Solution {
public:
    bool winnerSquareGame(int n) {
     dp[0] = false;
     for (int i = 1; i <= n; i++){
         dp[i] = false;
         for (int j = 1; j*j <= i; j++){
             if (!dp[i-j*j]){
                 dp[i] = true;
                 break;
             }
         }
     }
     return dp[n];
    }
};

// const int maxn = 1e5+5;

// bitset<maxn> dp;

// class Solution {
// public:
//     bool winnerSquareGame(int n) {
//         dp[0] = 0;
//         for (int i = 1; i <= n; i++){
//             dp[i] = 0;
//             for (int j = 1; j*j <= i; j++){
//                 if (!dp[i-j*j]){
//                     dp[i] = 1;
//                     break;
//                 }
//             }
//         }
//         return dp[n];
//     }
// };